transports/webtransport/src/transport.rs

```diff
@@ -0,0 +1,682 @@
+use std::{
+    collections::HashSet,
+    fmt,
+    future::Pending,
+    io,
+    net::{IpAddr, SocketAddr, UdpSocket},
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll, Waker},
+    time::Duration,
+};
+
+use futures::{future::BoxFuture, prelude::*, ready, stream::SelectAll};
+use if_watch::{tokio::IfWatcher, IfEvent};
+use libp2p_core::{
+    multiaddr::Protocol,
+    transport::{DialOpts, ListenerId, TransportError, TransportEvent},
+    Multiaddr, Transport,
+};
+use libp2p_identity::{Keypair, PeerId};
+use socket2::{Domain, Socket, Type};
+use wtransport::{
+    endpoint::{endpoint_side::Server, Endpoint, SessionRequest},
+    error::ConnectionError,
+    ServerConfig,
+};
+
+use crate::{certificate::CertHash, config::Config, connection::Connection, Connecting, Error};
+
+pub struct GenTransport {
+    config: Config,
+
+    listeners: SelectAll<Listener>,
+    /// Waker to poll the transport again when a new listener is added.
+    waker: Option<Waker>,
+}
+
+impl GenTransport {
+    pub fn new(config: Config) -> Self {
+        GenTransport {
+            config,
+            listeners: SelectAll::new(),
+            waker: None,
+        }
+    }
+
+    /// Extract the addr, quic version and peer id from the given [`Multiaddr`].
+    fn remote_multiaddr_to_socketaddr(
+        &self,
+        addr: Multiaddr,
+        check_unspecified_addr: bool,
+    ) -> Result<(SocketAddr, Option<PeerId>), TransportError<<Self as Transport>::Error>> {
+        let (socket_addr, peer_id) = multiaddr_to_socketaddr(&addr)
+            .ok_or_else(|| TransportError::MultiaddrNotSupported(addr.clone()))?;
+        if check_unspecified_addr && (socket_addr.port() == 0 || socket_addr.ip().is_unspecified())
+        {
+            return Err(TransportError::MultiaddrNotSupported(addr));
+        }
+        Ok((socket_addr, peer_id))
+    }
```

Comment:
We don't use the `check_unspecified_addr` option, so I don't think we don't need this method and can instead directly call `multiaddr_to_socketaddr` below in `listen_on`?